### ISSUE TITLE
Problems running on Windows

### ISSUE THREAD

**Issue by smillerk2** [binoculars/aws-lambda-ffmpeg#11]
I have downloaded ffmpeg-git-64bit-static.tar.xz and unpacked in the Build directory and see the ffmpeg and ffprobe files. When I run "npm install" using a Git bash window I receive the following error:

```
$ npm install

> aws-lambda-ffmpeg@0.0.0 postinstall c:\www\aws-lambda-ffmpeg
> node_modules/.bin/gulp postinstall

'node_modules' is not recognized as an internal or external command,
operable program or batch file.

npm ERR! aws-lambda-ffmpeg@0.0.0 postinstall: `node_modules/.bin/gulp postinstal
l`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the aws-lambda-ffmpeg@0.0.0 postinstall script.
npm ERR! This is most likely a problem with the aws-lambda-ffmpeg package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node_modules/.bin/gulp postinstall
npm ERR! You can get their info via:
npm ERR!     npm owner ls aws-lambda-ffmpeg
npm ERR! There is likely additional logging output above.
npm ERR! System Windows_NT 6.1.7601
npm ERR! command "c:\\Program Files (x86)\\nodejs\\node.exe" "c:\\Program Files
(x86)\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! cwd c:\www\aws-lambda-ffmpeg
npm ERR! node -v v0.10.22
npm ERR! npm -v 1.3.14
npm ERR! code ELIFECYCLE
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     c:\www\aws-lambda-ffmpeg\npm-debug.log
npm ERR! not ok code 0
```
The contents of the Build directory have been removed at this point and no dist folder has been created. Any ideas as to what is occurring?

**Comment by binoculars** [comment]
I think Git bash is cygwin-based, right? Check out [this post](http://soyuka.me/using-nodejs-with-cygwin-v0-10-25/). I had no issues installing via the node.js command prompt.

![image](https://cloud.githubusercontent.com/assets/3374510/11637213/2136a072-9cee-11e5-8bd4-89576bb6e1af.png)

```
H:\aws-lambda-ffmpeg>npm install
npm WARN deprecated aws-sdk@2.2.3: This version introduced a breaking change in
the way the CopySource parameter is passed in to Amazon S3. Please upgrade to v2
.2.4 or higher
npm WARN deprecated lodash@1.0.2: lodash@<3.0.0 is no longer maintained. Upgrade
 to lodash@^3.0.0
|
> aws-lambda-ffmpeg@0.0.0 postinstall H:\aws-lambda-ffmpeg
> gulp postinstall

[14:17:22] Using gulpfile H:\aws-lambda-ffmpeg\gulpfile.js
[14:17:22] Starting 'postinstall'...
[14:17:22] Finished 'postinstall' after 15 ms
gulp-rename@1.2.2 node_modules\gulp-rename

async@0.9.2 node_modules\async

uuid@2.0.1 node_modules\uuid

run-sequence@1.1.5 node_modules\run-sequence
└── chalk@1.1.1 (supports-color@2.0.0, escape-string-regexp@1.0.3, ansi-styles@2
.1.0, has-ansi@2.0.0, strip-ansi@3.0.0)

gulp-flatten@0.0.4 node_modules\gulp-flatten
└── through2@0.6.5 (xtend@4.0.1, readable-stream@1.0.33)

gulp-zip@3.0.2 node_modules\gulp-zip
├── yazl@2.2.2 (buffer-crc32@0.2.5)
├── concat-stream@1.5.1 (inherits@2.0.1, typedarray@0.0.6, readable-stream@2.0.4
)
├── chalk@1.1.1 (escape-string-regexp@1.0.3, supports-color@2.0.0, ansi-styles@2
.1.0, has-ansi@2.0.0, strip-ansi@3.0.0)
└── through2@0.6.5 (xtend@4.0.1, readable-stream@1.0.33)

del@1.2.1 node_modules\del
├── is-path-cwd@1.0.0
├── object-assign@3.0.0
├── each-async@1.1.1 (onetime@1.0.0, set-immediate-shim@1.0.1)
├── is-path-in-cwd@1.0.0 (is-path-inside@1.0.0)
├── rimraf@2.4.4 (glob@5.0.15)
└── globby@2.1.0 (array-union@1.0.1, async@1.5.0, glob@5.0.15)

gulp-install@0.4.0 node_modules\gulp-install
├── which@1.2.0 (is-absolute@0.1.7)
└── through2@0.6.5 (xtend@4.0.1, readable-stream@1.0.33)

gulp-util@3.0.7 node_modules\gulp-util
├── array-differ@1.0.0
├── array-uniq@1.0.2
├── lodash._reinterpolate@3.0.0
├── beeper@1.1.0
├── fancy-log@1.1.0
├── lodash._reescape@3.0.0
├── lodash._reevaluate@3.0.0
├── object-assign@3.0.0
├── replace-ext@0.0.1
├── has-gulplog@0.1.0 (sparkles@1.0.0)
├── minimist@1.2.0
├── gulplog@1.0.0 (glogg@1.0.0)
├── vinyl@0.5.3 (clone-stats@0.0.1, clone@1.0.2)
├── lodash.template@3.6.2 (lodash._basecopy@3.0.1, lodash._basetostring@3.0.1, l
odash._basevalues@3.0.0, lodash.escape@3.0.0, lodash.templatesettings@3.1.0, lod
ash._isiterateecall@3.0.9, lodash.restparam@3.6.1, lodash.keys@3.1.2)
├── chalk@1.1.1 (escape-string-regexp@1.0.3, supports-color@2.0.0, ansi-styles@2
.1.0, strip-ansi@3.0.0, has-ansi@2.0.0)
├── multipipe@0.1.2 (duplexer2@0.0.2)
├── through2@2.0.0 (xtend@4.0.1, readable-stream@2.0.4)
└── dateformat@1.0.12 (get-stdin@4.0.1, meow@3.6.0)

gulp@3.9.0 node_modules\gulp
├── interpret@0.6.6
├── pretty-hrtime@1.0.1
├── deprecated@0.0.1
├── archy@1.0.0
├── tildify@1.1.2 (os-homedir@1.0.1)
├── v8flags@2.0.10 (user-home@1.1.1)
├── semver@4.3.6
├── minimist@1.2.0
├── orchestrator@0.3.7 (stream-consume@0.1.0, sequencify@0.0.7, end-of-stream@0.
1.5)
├── chalk@1.1.1 (ansi-styles@2.1.0, supports-color@2.0.0, escape-string-regexp@1
.0.3, strip-ansi@3.0.0, has-ansi@2.0.0)
├── liftoff@2.2.0 (extend@2.0.1, rechoir@0.6.2, flagged-respawn@0.3.1, findup-sy
nc@0.3.0, resolve@1.1.6)
└── vinyl-fs@0.3.14 (graceful-fs@3.0.8, defaults@1.0.3, strip-bom@1.0.0, vinyl@0
.4.6, mkdirp@0.5.1, through2@0.6.5, glob-stream@3.1.18, glob-watcher@0.0.6)

gulp-shell@0.4.3 node_modules\gulp-shell
├── async@1.4.2
├── through2@2.0.0 (xtend@4.0.1, readable-stream@2.0.4)
└── lodash@3.10.1

aws-sdk@2.2.3 node_modules\aws-sdk
├── xmlbuilder@0.4.2
├── xml2js@0.2.8
└── sax@0.5.3

H:\aws-lambda-ffmpeg>
```

**Comment by smillerk2** [comment]
Thanks for your quick response - much appreciated. OK, I have updated my version of node and tried using the node.js command prompt. I only see a fraction of the output you reported:

```
C:\www\aws-lambda-ffmpeg>npm install

> aws-lambda-ffmpeg@0.0.0 postinstall C:\www\aws-lambda-ffmpeg
> gulp postinstall

[14:58:49] Using gulpfile C:\www\aws-lambda-ffmpeg\gulpfile.js
[14:58:49] Starting 'postinstall'...
[14:58:49] Finished 'postinstall' after 1.39 ms
```
No dist folder has been created and I don't see any other errors.

```
C:\www\aws-lambda-ffmpeg>npm -v
3.3.12

C:\www\aws-lambda-ffmpeg>node -v
v5.1.1
```

**Comment by binoculars** [comment]
No prob. I'm using 0.10.36 (because that's what the lambda runtime is), but as long as you stick to compatibility with that version, it should still work.

You'll need to run gulp for the default task (just `gulp` from the CLI). That will set everything up. Although, I doubt it'll work well on windows without tar--maybe with babun. I tried a number of things to natively extract the xz ffmpeg package, but none of them worked.

**Comment by smillerk2** [comment]
I have been able to get it to work (thank you).

```
npm init
npm install -g gulp
gulp
```
I commented out the download-ffmpeg and untar-ffmpeg tasks. The dist.zip file has been created and uploaded to AWS. Next task is figuring out why it doesn't complete in AWS even with max memory and a high timeout:

```
2015-12-07T20:33:22.780Z	c24c73ef-9d21-11e5-aa4e-41551ebefbb1	Starting download
2015-12-07T20:33:23.052Z	c24c73ef-9d21-11e5-aa4e-41551ebefbb1	starting ffprobe
Process exited before completing request
```

**Comment by binoculars** [comment]
were you able to manually untar FFmpeg and FFprobe?

**Comment by smillerk2** [comment]
Yes, build folder contains:

```
ffmpeg-git-20151206-64bit-static
ffmpeg-git-20151206-64bit-static\ffmpeg
ffmpeg-git-20151206-64bit-static\ffprobe
ffmpeg-git-20151206-64bit-static\ffmpeg-10bit
ffmpeg-git-20151206-64bit-static\ffserver

After running gulp, those files vanish and are presumably bundled somewhere in dist.zip.

**Comment by binoculars** [comment]
They vanish because the `clean` task deletes them. You probably want to go ahead nd comment that out as well.

**Comment by smillerk2** [comment]
OK. Commented out that section and also changed my version of node to v0.10.36 to match AWS. ffprobe doesn't seem to complete though. Does the source file name have to be a UUID? I'm not sure what else could be causing the Lamdba to terminate unexpectedly. Nothing in the AWS logs.

```
2015-12-08T14:56:34.675Z	df246e9f-9dbb-11e5-bb8c-b7fe663b9bee	srcKey: video/2015-12-03-17-00-30-01.avi
2015-12-08T14:56:34.675Z	df246e9f-9dbb-11e5-bb8c-b7fe663b9bee	keyPrefix: video/2015-12-03-17-00-30-01
2015-12-08T14:56:34.675Z	df246e9f-9dbb-11e5-bb8c-b7fe663b9bee	id: <Buffer 20 15 12 03 17 00 30 01 00 00 00 00 00 00 00 00>
2015-12-08T14:56:34.676Z	df246e9f-9dbb-11e5-bb8c-b7fe663b9bee	Starting download
2015-12-08T14:56:34.816Z	df246e9f-9dbb-11e5-bb8c-b7fe663b9bee	starting ffprobe
Process exited before completing request
```

**Comment by binoculars** [comment]
it doesn't have to be a UUID. looks like FFprobe is crashing. Are you sure it's included in dist.zip ?

**Comment by smillerk2** [comment]
Positive. I've extracted dist.zip (30Mb) to a temp folder and here are the contents:

```
$ ls -a -R
.:
.  ..  config.json  ffmpeg  ffprobe  index.js  node_modules

./node_modules:
.  ..  async  uuid

./node_modules/async:
.   LICENSE    bower.json      lib           support
..  README.md  component.json  package.json

./node_modules/async/lib:
.  ..  async.js

./node_modules/async/support:
.  ..  sync-package-managers.js

./node_modules/uuid:
.   LICENSE.md  benchmark  package.json    rng.js  uuid.js
..  README.md   misc       rng-browser.js  test

./node_modules/uuid/benchmark:
.   README.md  bench.sh            benchmark.js
..  bench.gnu  benchmark-native.c  package.json

./node_modules/uuid/misc:
.  ..  compare.js  perf.js

./node_modules/uuid/test:
.  ..  mocha.opts  test.js
``

**Comment by binoculars** [comment]
It looks like you have everything you need. Usually when something like this happens, I try to test it locally. Configure `test_event.json` and run `npm test` (or just `node index.js`)

**Comment by binoculars** [comment]
I'm trying myself from a windows box. I'm getting the same error. I have no idea what's causing it right now (thinking it might be an STDIN error), but I intend to find out tonight.

**Comment by smillerk2** [comment]
Thanks for investigating. Running npm test also failed locally but with a unhandled error event:

```
video/2015-12-03-17-00-30-01.avi
video/2015-12-03-17-00-30-01
<Buffer 20 15 12 03 17 00 30 01 00 00 00 00 00 00 00 00>
Starting download

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: EISDIR, open 'C:\tmp\download'
```
The directory does exist.

**Comment by binoculars** [comment]
I did some work last night to *try* to fix the issue. It works fine when I run it from my Mac, but when I come back to windows and try to upload the function, I get a weird error.

```
2015-12-09T15:01:34.862Z	bc1ffceb-9e85-11e5-a060-7741f65af2c5	Starting download
2015-12-09T15:01:35.245Z	bc1ffceb-9e85-11e5-a060-7741f65af2c5	Starting FFprobe
2015-12-09T15:01:35.253Z	bc1ffceb-9e85-11e5-a060-7741f65af2c5	{"errorMessage":"spawn EACCES","errorType":"Error","stackTrace":["errnoException (child_process.js:1011:11)","Process.ChildProcess._handle.onexit (child_process.js:802:34)"]}
END RequestId: bc1ffceb-9e85-11e5-a060-7741f65af2c5
REPORT RequestId: bc1ffceb-9e85-11e5-a060-7741f65af2c5	Duration: 396.49 ms	Billed Duration: 400 ms 	Memory Size: 1024 MB	Max Memory Used: 49 MB	
```

At this point, I'm out of ideas. The best recommendation I can make would be to try it in a linux VM or Docker. If you know more about Node on Windows and can figure it out, please send a PR.

**Comment by binoculars** [comment]
Closing. The initial problem was resolved and the permissions issue is addressed in #13
